mani_skill/envs/tasks/rotate_cube.py

```diff
@@ -0,0 +1,438 @@
+            obs.update(
+                obj_p=self.obj.pose.p,
+                obj_q=self.obj.pose.q,
+            )
+            pass
+        return obs
+
+    def compute_dense_reward(self, obs: Any, action: Array, info: Dict):
+        obj_pos = self.obj.pose.p
+        obj_q = self.obj.pose.q
+        goal_pos = self.obj_goal.pose.p
+        goal_q = self.obj_goal.pose.q
+
+        object_dist_weight = 5
+        object_rot_weight = 5
+
+        # Reward penalising finger movement
+
+        tip_poses = self.agent.tip_poses()
+        # shape (N, 3 + 4, 3 fingers)
+
+        finger_reach_object_dist_1 = torch.norm(tip_poses[:, :3, 0] - obj_pos, p=2, dim=-1)
+        finger_reach_object_dist_2 = torch.norm(tip_poses[:, :3, 1] - obj_pos, p=2, dim=-1)
+        finger_reach_object_dist_3 = torch.norm(tip_poses[:, :3, 2] - obj_pos, p=2, dim=-1)
+        finger_reach_object_reward1 = 1 - torch.tanh(5 * finger_reach_object_dist_1)
+        finger_reach_object_reward2 = 1 - torch.tanh(5 * finger_reach_object_dist_2)
+        finger_reach_object_reward3 = 1 - torch.tanh(5 * finger_reach_object_dist_3)
+        finger_reach_object_reward = object_dist_weight * (finger_reach_object_reward1 + finger_reach_object_reward2 + finger_reach_object_reward3) / 3
+
+        # Reward for object distance
+        object_dist = torch.norm(obj_pos - goal_pos, p=2, dim=-1)
+
+        init_xyz_tensor = torch.tensor([0, 0, 0.032], dtype=torch.float, device=self.device).reshape(1, 3)
+        init_z_dist = torch.norm(init_xyz_tensor - goal_pos[...,], p=2, dim=-1)
+
+        # object_dist_reward = object_dist_weight * dt * lgsk_kernel(object_dist, scale=50., eps=2.)
+
+        object_dist_reward = 1 - torch.tanh(5 * object_dist)
+        object_init_dist_reward = 1 - torch.tanh(5 * init_z_dist)
+        object_dist_reward -= object_init_dist_reward
+
+        init_z_tensor = torch.tensor([0.032], dtype=torch.float, device=self.device).reshape(1, 1)
+        object_z_dist = torch.norm(obj_pos[..., 2:3] - goal_pos[..., 2:3], p=2, dim=-1)
+        init_z_dist = torch.norm(init_z_tensor - goal_pos[..., 2:3], p=2, dim=-1)
+        object_lift_reward = 5 * ((1 - torch.tanh(5 * object_z_dist)))
+        object_init_z_reward = 5 * ((1 - torch.tanh(5 * init_z_dist)))
+
+        object_lift_reward -= object_init_z_reward
+
+        # extract quaternion orientation
+        angles = quat_diff_rad(obj_q, goal_q)
+        object_rot_reward = -1 * torch.abs(angles)
+        pose_reward = object_dist_weight * (object_dist_reward + object_lift_reward) + object_rot_weight * object_rot_reward
+        total_reward = finger_reach_object_reward + pose_reward
+        total_reward[info["success"]] = 3
+        return total_reward
+
+    def compute_normalized_dense_reward(self, obs: Any, action: Array, info: Dict):
+        return self.compute_dense_reward(obs=obs, action=action, info=info) / 1
+
+
+@torch.jit.script
+def quat_from_euler_xyz(roll, pitch, yaw):
```

Comment:
our other code is using snippets from pytorch3d for rotation conversions. Is this a fast implementation compared to that one? If so we should move this code to the mani_skill/utils/geometry/rotation_conversions.py file

Reply:
Thanks for reminding me. I just forgot about this library.